addons/pkg-maven/common/src/main/java/org/commonjava/indy/pkg/maven/content/MetadataMergeListner.java

```diff
@@ -0,0 +1,207 @@
+    @Inject
+    private StoreDataManager storeManager;
+
+    @Inject
+    @MetadataCache
+    private CacheHandle<StoreKey, Map> metadataCache;
+
+    /**
+     * Listen to an #{@link ArtifactStorePreUpdateEvent} and clear the metadata cache due to changed memeber in that event
+     *
+     * @param event
+     */
+    public void onStoreUpdate( @Observes final ArtifactStorePreUpdateEvent event )
+    {
+
+        logger.trace( "Got store-update event: {}", event );
+
+        if ( ArtifactStoreUpdateType.UPDATE == event.getType() )
+        {
+            for ( ArtifactStore store : event )
+            {
+                removeMetadataCacheContent( store, event.getChangeMap() );
+            }
+        }
+    }
+
+    private void removeMetadataCacheContent( final ArtifactStore store,
+                                             final Map<ArtifactStore, ArtifactStore> changeMap )
+    {
+        handleStoreDisableOrEnable( store, changeMap );
+
+        handleGroupMembersChanged( store, changeMap );
+    }
+
+    // if a store is disabled/enabled, we should clear its metadata cache and all of its affected groups cache too.
+    private void handleStoreDisableOrEnable(final ArtifactStore store,
+                                            final Map<ArtifactStore, ArtifactStore> changeMap){
+        final ArtifactStore oldStore = changeMap.get( store );
+        if ( store.isDisabled() != oldStore.isDisabled() )
+        {
+            final Map<String, Metadata> metadataMap = metadataCache.get( store.getKey() );
+            if ( metadataMap != null && !metadataMap.isEmpty() )
+            {
+                metadataCache.remove( store.getKey() );
+                storeManager.getGroupsAffectedBy( store.getKey() ).forEach( g -> clearGroupMetaCache( g ) );
+            }
+        }
+    }
+
+    // If group members changed, should clear the cascading groups metadata cache
+    private void handleGroupMembersChanged(final ArtifactStore store,
+                                           final Map<ArtifactStore, ArtifactStore> changeMap){
+        final StoreKey key = store.getKey();
+        if ( StoreType.group == key.getType() )
+        {
+            final List<StoreKey> newMembers = ( (Group) store ).getConstituents();
+            logger.trace( "New members of: {} are: {}", store.getKey(), newMembers );
+
+            final Group group = (Group) changeMap.get( store );
+            final List<StoreKey> oldMembers = group.getConstituents();
+            logger.trace( "Old members of: {} are: {}", group.getName(), oldMembers );
+
+            boolean membersChanged = false;
+
+            if ( newMembers.size() != oldMembers.size() )
+            {
+                membersChanged = true;
+            }
+            else
+            {
+                for ( StoreKey storeKey : newMembers )
+                {
+                    if ( !oldMembers.contains( storeKey ) )
+                    {
+                        membersChanged = false;
```

Comment:
In this else block, there is no way to set `membersChanged = true`, and I think there's no way this assignment will change the state of that field. Is that intended?

Reply:
Oops. This is a wrong set, it should be membersChanged=true as it is a member change which a new member added(old does not contain). What I'm thinking at that time...